osu.Game/Database/StandardisedScoreMigrationTools.cs

```diff
@@ -461,6 +482,94 @@ private static long convertFromLegacyTotalScore(ScoreInfo score, LegacyBeatmapCo
             return convertedTotalScore;
         }
 
+        /// <summary>
+        /// <para>
+        /// For catch, the general method of calculating the combo proportion used for other rulesets is generally useless.
+        /// This is because in stable score V1, catch has quadratic score progression,
+        /// while in stable score V2, score progression is logarithmic up to 200 combo and then linear.
+        /// </para>
+        /// <para>
+        /// This means that applying the naive rescale method to scores with lots of short combos (think 10x 100-long combos on a 1000-object map)
+        /// by linearly rescaling the combo portion as given by score V1 leads to horribly underestimating it.
+        /// Therefore this method attempts to counteract this by calculating the best case estimate for the combo proportion that takes all of the above into account.
+        /// </para>
+        /// <para>
+        /// The general idea is that aside from the <paramref name="scoreMaxCombo"/> which the player is known to have hit,
+        /// the remaining misses are evenly distributed across the rest of the objects that give combo.
+        /// This is therefore a worst-case estimate.
+        /// </para>
+        /// </summary>
+        private static double estimateComboProportionForCatch(int beatmapMaxCombo, int scoreMaxCombo, int scoreMissCount)
+        {
+            if (beatmapMaxCombo == 0)
+                return 1;
+
+            if (scoreMaxCombo == 0)
+                return 0;
+
+            if (beatmapMaxCombo == scoreMaxCombo)
+                return 1;
+
+            double estimatedBestCaseTotal = estimateBestCaseComboTotal(beatmapMaxCombo);
+
+            int remainingCombo = beatmapMaxCombo - (scoreMaxCombo + scoreMissCount);
+            double totalDroppedScore = 0;
+
+            int assumedLengthOfRemainingCombos = (int)Math.Floor((double)remainingCombo / scoreMissCount);
+
+            if (assumedLengthOfRemainingCombos > 0)
+            {
+                while (remainingCombo > 0)
+                {
+                    int comboLength = Math.Min(assumedLengthOfRemainingCombos, remainingCombo);
+
+                    remainingCombo -= comboLength;
+                    totalDroppedScore += estimateDroppedComboScoreAfterMiss(comboLength);
+                }
+            }
```

Comment:
As discussed on Discord, you can remove the while loop here, if that even matters.
```suggestion
           if (assumedLengthOfRemainingCombos > 0)
        {
            // How many combos we have to evaluate (still estimate)
            int combosCount = (int)Math.Floor((double)remainingCombo / assumedLengthOfRemainingCombos);
            
            // We may still have objects remaining after the last combo evaluation, if remaining objects can't be evenly split by the assumed length of remaining combos
            int remainingObjects = remainingCombo - (combosCount * assumedLengthOfRemainingCombos);

            totalDroppedScore += combosCount * estimateDroppedComboScoreAfterMiss(assumedLengthOfRemainingCombos);
            
            if (remainingObjects > 0)
            {
                totalDroppedScore += estimateDroppedComboScoreAfterMiss(remainingObjects);
            }
        }
```

Reply:
Yeah I'll gauge the general reaction to *vaguely points at everything else in this diff* and if it doesn't get kicked to the curb I'll apply this change and re-validate that results don't change.

Reply:
Have applied.

!diffcalc
GENERATORS=score
RULESET=catch